Add Pybindings for Program.h/cpp

Conarnar opened the pull request:
### Summary

Today our python apis in executorch.runtime are implemented off of extension/pybindings which only offers a module api. We would like to migrate to having the lower level ET api exposed to python directly and then writing the module api in python. The first step to this is adding pybindings for Program.

Bindings for the class Program and its methods num_methods and get_method_name were added.

### Test plan

Tests were added to `extension/pybindings/test/make_test.py`
1. test_program_methods_one -- verifies num_methods and get_method_name works with one method
2. test_program_methods_multi -- verifies num_methods and get_method_name works with multiple methods
3. test_program_method_index_out_of_bounds -- verifies get_method_name raises a runtime error if index is out of bounds

Conarnar commented:
@pytorchbot label "release notes: none"

lucylq commented on extension/pybindings/pybindings.cpp:
```diff
@@ -965,6 +965,99 @@ struct PyModule final {
+    return std::string(res.get());
+  }
+
+  std::string get_output_flattening_encoding(std::string method_name) const {
```
This API is deprecated, should it still be exposed? Maybe mark it as deprecated as well. 

https://github.com/pytorch/executorch/blob/7e28a044d7acf8ae405d9d5a37d11ceb12b90899/runtime/executor/program.h#L168

cc @JacobSzwejbka

Conarnar committed:
Add Pybindings for Program.h/cpp (#12016)

Summary:
Today our python apis in executorch.runtime are implemented off of extension/pybindings which only offers a module api. We would like to migrate to having the lower level ET api exposed to python directly and then writing the module api in python. The first step to this is adding pybindings for Program.

Bindings for the class Program and its methods num_methods and get_method_name were added.


Test Plan:
Tests were added to `extension/pybindings/test/make_test.py`
1. test_program_methods_one -- verifies num_methods and get_method_name works with one method
2. test_program_methods_multi -- verifies num_methods and get_method_name works with multiple methods
3. test_program_method_index_out_of_bounds -- verifies get_method_name raises a runtime error if index is out of bounds

Rollback Plan:

Reviewed By: JacobSzwejbka

Differential Revision: D77388495

Pulled By: Conarnar
